pageserver/src/pgdatadir_mapping.rs

```diff
@@ -154,7 +165,18 @@ impl<R: Repository> DatadirTimeline<R> {
 
         let key = rel_size_to_key(tag);
         let mut buf = self.tline.get(key, lsn)?;
-        Ok(buf.get_u32_le())
+        let nblocks = buf.get_u32_le();
+
+        // Update relation size cache
+        let mut relsize_cache = self.relsize_cache.write().unwrap();
+        if let Some((cached_lsn, _cached_nblocks)) = relsize_cache.get(&tag) {
+            if lsn >= *cached_lsn {
+                relsize_cache.insert(tag, (lsn, nblocks));
+            }
+        } else {
+            relsize_cache.insert(tag, (lsn, nblocks));
+        }
```

Comment:
Can be used with [`Hashmap::entry`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.entry) to avoid additional lookup

Reply:
Done